juju/juju
# feat(units): add support for updating caas units

The caas application provisioner work now updates caas units in dqlite. This includes setting the unit password.

Part of the work was to refactor some of the early implementation for create application so that the application creation and addition of units are separate state operations in an atomic transaction. We also distinguish between create and import of applications, and adding a unit at deploy time vs inserting a unit during import or when registering a caas unit. The latter requires that password and all other unit attributes can be set; the former just needs to unsert a minimal unit record.

## QA steps

test on both k8s and vm

juju bootstrap microk8s c1
juju add-model foo
juju deploy snappass-test -n 2
juju bootstrap microk8s c2
juju switch c1
juju migrate foo c2

juju bootstrap lxd c1
juju add-model foo
juju deploy ubuntu -n 2
juju bootstrap lxd c2
juju switch c1
juju migrate foo c2

On the lxd model, run the dqlite repl
select from the unit_agent_status and unit_workload_status tables to check status rows have been written

## Links

**Jira card:** [JUJU-6765](https://warthogs.atlassian.net/browse/JUJU-6765)
